modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java

```diff
@@ -0,0 +1,273 @@
+                int partitionId,
+                @Nullable RowId nextRowIdToBuiltFromPreviousBatch
+        ) {
+            this.table = table;
+            this.tableIndexView = tableIndexView;
+            this.partitionId = partitionId;
+            this.nextRowIdToBuiltFromPreviousBatch = nextRowIdToBuiltFromPreviousBatch;
+        }
+
+        @Override
+        public void run() {
+            if (!busyLock.enterBusy()) {
+                return;
+            }
+
+            try {
+                // At the time of creating the index, we should have already waited for the table to be created and its raft of clients
+                // (services) to start for all partitions, so there should be no errors.
+                RaftGroupService raftGroupService = table.internalTable().partitionRaftGroupService(partitionId);
+
+                raftGroupService
+                        // We do not check the presence of nodes in the topology on purpose, so as not to get into races on
+                        // rebalancing, it will be more convenient and reliable for us to wait for a stable topology with a chosen
+                        // leader.
+                        .refreshAndGetLeaderWithTerm()
+                        .thenComposeAsync(leaderWithTerm -> {
+                            if (!busyLock.enterBusy()) {
+                                return completedFuture(null);
+                            }
+
+                            try {
+                                // At this point, we have a stable topology, each node of which has already applied all local updates.
+                                if (!localNodeConsistentId().equals(leaderWithTerm.leader().consistentId())) {
+                                    // TODO: IGNITE-19053 Must handle the change of leader
+                                    // TODO: IGNITE-19053 Add a test to change the leader even at the start of the task
+                                    return completedFuture(null);
+                                }
+
+                                List<RowId> batchRowIds = collectRowIdBatch();
+
+                                RowId nextRowId = getNextRowIdForNextBatch(batchRowIds);
+
+                                boolean finish = batchRowIds.size() < BUILD_INDEX_ROW_ID_BATCH_SIZE || nextRowId == null;
+
+                                // TODO: IGNITE-19053 Must handle the change of leader
+                                return raftGroupService.run(createBuildIndexCommand(batchRowIds, finish))
+                                        .thenRun(() -> {
+                                            if (!finish) {
+                                                assert nextRowId != null : createCommonTableIndexInfo();
+
+                                                buildIndexExecutor.submit(
+                                                        new BuildIndexTask(table, tableIndexView, partitionId, nextRowId)
```

Comment:
Can we pass `this` as the argument?

Reply:
We can, but why?

Reply:
Because I don't understand the need to create a new object, if we already have one

Reply:
What's wrong with creating fresh objects?
GC will remove old objects, will not shift to old generations.